File: lib/utils/rotation_engine.dart
class RotationEngine {
  /// Generates the circular card passing logic for Sabotage rounds.
  /// Returns a map of Rotation Round (1 to sabotageRounds) -> Map of `holdingPlayerId` -> `targetPlayerId`
  /// 
  /// The algorithm ensures:
  /// 1. No player receives their own card (Target != Holder).
  /// 2. No player receives the same Target's card structurally.
  static Map<int, Map<String, String>> generateRotations(List<String> playerIds, int sabotageRounds) {
    if (playerIds.length <= sabotageRounds) {
      throw Exception('Total players must be strictly greater than sabotage rounds to prevent players from receiving their own cards.');
    }
    
    var rotations = <int, Map<String, String>>{};
    int p = playerIds.length;
    
    // Create a local copy to ensure predictable indexing logic without destroying the original list
    List<String> orderedIds = List.from(playerIds);
    // By keeping the list explicitly unshuffled during assignment generation, 
    // the mathematical modulo ensures perfect distribution. 
    // Randomization of prompts themselves will handle variety.
    
    for (int r = 1; r <= sabotageRounds; r++) {
      Map<String, String> currentRoundAssignments = {};
      for (int i = 0; i < p; i++) {
        // Derangement logic: Offset by the current rotation index
        int targetIndex = (i + r) % p;
        currentRoundAssignments[orderedIds[i]] = orderedIds[targetIndex];
      }
      rotations[r] = currentRoundAssignments;
    }
    
    return rotations;
  }
}

File: functions/src/rotation_engine.ts
export class RotationEngine {
  /**
   * Generates the circular card passing logic for Sabotage rounds.
   * Returns a map of Rotation Round (1 to forgeriesPerCard) -> Map of holdingPlayerId -> targetPlayerId
   */
  static generateRotations(playerIds: string[], forgeriesPerCard: number): Record<number, Record<string, string>> {
    if (!Number.isInteger(forgeriesPerCard) || forgeriesPerCard < 1) {
      throw new Error(`forgeriesPerCard must be a positive integer, received: ${JSON.stringify(forgeriesPerCard)}`);
    }
    if (playerIds.length <= forgeriesPerCard) {
      throw new Error('Total players must be strictly greater than forgeries per card to prevent players from receiving their own cards.');
    }
    
    const rotations: Record<number, Record<string, string>> = {};
    const p = playerIds.length;
    const orderedIds = [...playerIds];
    
    for (let r = 1; r <= forgeriesPerCard; r++) {
      const currentRoundAssignments: Record<string, string> = {};
      for (let i = 0; i < p; i++) {
        const targetIndex = (i + r) % p;
        currentRoundAssignments[orderedIds[i]] = orderedIds[targetIndex];
      }
      rotations[r] = currentRoundAssignments;
    }
    
    return rotations;
  }
}
